src/main/java/software/amazon/smithy/gradle/SmithyPlugin.java

```diff
@@ -39,49 +40,35 @@
     private static final String DEFAULT_CLI_VERSION = "0.9.4";
     private static final List<String> SOURCE_DIRS = ListUtils.of(
             "model", "src/$name/smithy", "src/$name/resources/META-INF/smithy");
-
-    private boolean appliedPlugin;
+    private static final Logger LOGGER = Logger.getLogger(SmithyPlugin.class.getName());
 
     @Override
     public void apply(Project project) {
+        // Ensure that the Java plugin is applied.
         project.getPluginManager().apply(JavaPlugin.class);
-        project.getPluginManager().withPlugin("java", javaPlugin -> {
-            appliedPlugin = true;
-            SmithyExtension extension = project.getExtensions().create("smithy", SmithyExtension.class);
-            project.afterEvaluate(p -> {
-                registerSourceSets(project);
-                registerTasks(project, extension);
-            });
-        });
 
+        // Register the Smithy extension so that tasks can be configured.
+        project.getExtensions().create("smithy", SmithyExtension.class);
+
+        // Register the "smithyBuildJar" task. It's configured once the extension is available.
+        TaskProvider<SmithyBuildJar> provider = project.getTasks().register("smithyBuildJar", SmithyBuildJar.class);
+
+        // Can't read from the "smithy" extension until the "afterEvaluate" step.
         project.afterEvaluate(p -> {
-            if (!appliedPlugin) {
-                throw new GradleException("Smithy plugin is missing the required java plugin");
-            }
+            registerSourceSets(project);
+            addCliDependencies(project);
+            registerSmithyBuildTask(provider.get(), project);
         });
     }
 
-    private void registerTasks(Project project, SmithyExtension extension) {
-        SmithyBuildJar buildTask = project.getTasks().create("smithyBuild", SmithyBuildJar.class, task -> {
-            task.setProjection(extension.getProjection());
```

Comment:
All of this configuration was moved into the SmithyBuildJar task itself. We just use the extension for defaults and you can explicitly override configuration per/task. Values from the SmithyExtension are pull into tasks in the `@TaskAction` method (which, at that point, the extensions have all been resolved).